src/user_input.rs

```diff
@@ -1,444 +1,827 @@
+    DualGamepadAxis(DualGamepadAxisThreshold),
+    /// A button on a keyboard
+    Keyboard(KeyCode),
+    /// A button on a mouse
+    Mouse(MouseButton),
+}
+
+/// Used to indicate at which point a gamepad axis event should trigger an action.
+#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
+pub struct SingleGamepadAxisThreshold {
+    /// The axis that is being checked.
+    pub axis: GamepadAxisType,
+    /// Indicates whether the action should trigger when the axis value is greater, lesser, or equal
+    /// to the `threshold`.
+    pub comparison: GamepadAxisComparison,
+    /// The threshold to compare the current axis value with.
+    pub threshold: f32,
+}
+
+impl PartialEq for SingleGamepadAxisThreshold {
+    fn eq(&self, other: &Self) -> bool {
+        self.axis == other.axis
+            && self.comparison == other.comparison
+            && FloatOrd(self.threshold) == FloatOrd(other.threshold)
+    }
+}
+impl Eq for SingleGamepadAxisThreshold {}
+impl std::hash::Hash for SingleGamepadAxisThreshold {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.axis.hash(state);
+        self.comparison.hash(state);
+        FloatOrd(self.threshold).hash(state);
+    }
+}
+
+/// Used to indicate at which point a dual gamepad axis event should trigger an action.
+#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
+pub struct DualGamepadAxisThreshold {
+    /// The gamepad axis to use as the x axis
+    pub x_axis: GamepadAxisType,
+    /// The gamepad axis to use as the y axis
+    pub y_axis: GamepadAxisType,
+    /// Indicates whether the action should trigger when the axis movement is greater, lesser, or
+    /// equal to the `threshold`.
+    pub comparison: GamepadAxisComparison,
+    /// The threshold to compare the dual axis value with.
+    ///
+    /// The value of a dual axis is
+    pub threshold: f32,
+}
+
+impl PartialEq for DualGamepadAxisThreshold {
+    fn eq(&self, other: &Self) -> bool {
+        self.x_axis == other.x_axis
+            && self.y_axis == other.y_axis
+            && self.comparison == other.comparison
+            && FloatOrd(self.threshold) == FloatOrd(other.threshold)
+    }
+}
+impl Eq for DualGamepadAxisThreshold {}
+impl std::hash::Hash for DualGamepadAxisThreshold {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.x_axis.hash(state);
+        self.y_axis.hash(state);
+        self.comparison.hash(state);
+        FloatOrd(self.threshold).hash(state);
+    }
+}
+
+/// A virtual dpad that you will be able to read as an [`AxisPair`]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct VirtualDPad {
+    up: InputKind,
+    down: InputKind,
+    left: InputKind,
+    right: InputKind,
+}
+
+/// Different possible ways to compare an axis threshold to the axis value.
+///
+/// See [`GamepadAxisThreshold`].
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub enum GamepadAxisComparison {
```

Comment:
For `Equal` the use-case I could imagine was if you wanted an action to be triggered when the joystick was at `0.0`. It was easy to implement and not confusing to the user so I threw it in there, but I'm not sure if we really need it.

For `Always` the idea was if you needed to monitor an axis constantly, but you still didn't know which axis, because the user was allowed to configure the input mapping. I think that's a valid use-case. That way you can still query the axis value, similar to how you would monitor an axis in normal Bevy, but you still get the input abstraction where you don't have to know exactly which axis / virtual DPad / etc. triggered the input.

Otherwise you could emulate it by triggering less than `10000000`, or something like that, but that seems sloppy.

Still, I don't personally have a use-case for it yet, so maybe we leave it out unless somebody asks for it?

@CleanCut do you have any thoughts on this one?

Reply:
I just ran into the need for an `AbsoluteGreater` variant, to trigger if the axis movement was greater than a certain amount in any direction.

Do you think that `AbsoluteGreater` and `AbsoluteLess` variants make sense?

Reply:
> @CleanCut do you have any thoughts on this one?

I don't see a need for either `Equal` or `Always`.

> Do you think that AbsoluteGreater and AbsoluteLess variants make sense?

Given that `bevy_input` has a way to configure dead zones so that they're filtered out at that level AND the fact that `bevy_input` uses the term threshold to mean "the minimum difference between old and new values to apply the changes", I'm wondering if we should be managing deadzones at this level?  Maybe we don't need to? If we do, we should probably use the same terminology as `bevy_input`.

```
// the bevy_input way...
fn configure_gamepads(mut settings: ResMut<GamepadSettings>) {
    // add a larger default dead-zone to all axes (ignore small inputs, round to zero)
    settings.default_axis_settings.negative_low = -0.125;
    settings.default_axis_settings.positive_low = 0.125;

    // the axis value must change this much before it will register as a different value
    settings.default_axis_settings.threshold = 0.05;
}
```

Reply:
I like the absolute variants; feel free to add those.

Seconding @CleanCut on the "should we just be using bevy_inputs" settings though :)

Reply:
Would it be clearer to name the two structs `SingleGamepadAxis` and `DualGamepadAxis`? (Drop the `Threshold` suffix)

Reply:
> I'm wondering if we should be managing deadzones at this level? Maybe we don't need to? If we do, we should probably use the same terminology as bevy_input.

My reasoning behind managing deadzones at this level has to do with action specific dead-zones. For instance, in my game I have a character that I use the left stick to walk around with. For controlling the character, if you tilt the stick very slightly, I still want the `Move` action to be triggered and the player to move based on how far the stick is tilting.

But I also have a separate set of actions for the Menu, such as `Next`, and `Previous` which have bindings to the left stick as well. Since `Next` and `Previous` are not variable actions, I really only want to trigger them if the stick moves more than 50% in a direction, and I want that to be configured separate from the movement deadzone.

That is a good point about the term `threshold`. That could be confusing. I'm just not sure if we can use the same terminology as Bevy because we are trying to represent a different concept. In bevy, the settings say ( roughtly ), "at which point should I ignore the stick movement". With leafwing input we are saying, "at which point should my stick movement trigger an in-game response".

Another note: since this plugin is getting our input from bevy, the `bevy_input::GamepadSettings` are still applied, but they are applied before leafwing gets the values.

> Would it be clearer to name the two structs SingleGamepadAxis and DualGamepadAxis? (Drop the Threshold suffix)

Yeah, I like that.

---

Anyway, I'll take another look at it to see if I can come up with a more intuitive way to structure the enum maybe. Or let me know if you still think that maybe we should try to get rid of local thresholds in favor of a more global setting, or if you have any other ideas. :)

Reply:
I'm fine with the local thresholds :) That's not something I'd thought of before but I think it's actually the correct behavior.

Reply:
> That is a good point about the term threshold. That could be confusing. I'm just not sure if we can use the same terminology as Bevy because we are trying to represent a different concept.

Mirring Bevy's terminology is probably the clearest approach. I agree there is a subtle difference, but I think the semantics are essentially the same. In both cases, input is being omitted for a certain range of values.

My suggestion would be to split your current `threshold` field into the analogous `positive_low` and `negative_low` fields from Bevy's [`AxisSettings`](https://docs.rs/bevy/latest/bevy/input/gamepad/struct.AxisSettings.html).  Then, if we want to support more settings similar to `AxisSettings`, we can similarly add the same field names to your structs.

> Or let me know if you still think that maybe we should try to get rid of local thresholds in favor of a more global setting, or if you have any other ideas. :)

Local settings are fine. I wasn't really going for global over local settings, per se. I just used Bevy's global settings as my example.  

> Anyway, I'll take another look at it to see if I can come up with a more intuitive way to structure the enum maybe. 

I wonder if it would make sense to combine `SingleGamepadAxis` and `DualGamepadAxis` into a single type. But that may be a bridge too far for this PR. I haven't been able to come up with a good design for that in my head. (Just thinking out loud on this one--I'm not sure this is actionable.)

Reply:
> My suggestion would be to split your current threshold field into the analogous positive_low and negative_low

OK, I see, that makes sense. So if you wanted to trigger on any positive axis movement, you would set the `negative_low: -1.1` to make the negative movement never trigger, and the `positive_low: 0.0` to make positive move always trigger.

I think that makes sense and handles all the real-world use-cases I can think of. Good call!

I was just messing around with alternatives and came up with this, to get rid of `threshold`, but it's overkill and I like the bevy convention better. :) 

```rust
pub enum Comparison {
    /// Value must be less than or equal to a value
    Less(f32),
    /// Value must be greater than or equal to a value
    Greater(f32),
    /// The absolute value must be less than or equal to a value
    AbsoluteLess(f32),
    /// The absolute value must be greater than or equal to a value
    AbsoluteGreater(f32),
    /// The value must be within the given range
    WithinRange(f32, f32),
}
```

> I wonder if it would make sense to combine SingleGamepadAxis and DualGamepadAxis into a single type. 

Not a bad thought, but with all the ways I've thought of it so far I think it ends up cleaner to have them separate.

We could follow up if anything comes to mind.